File: code/utils/project_paths.py
from __future__ import annotations

import json
from pathlib import Path


def find_repo_root(start: Path | None = None) -> Path:
    start = Path.cwd().resolve() if start is None else Path(start).resolve()
    for candidate in [start, *start.parents]:
        if (candidate / "config" / "paths.json").exists():
            return candidate
    raise FileNotFoundError("Could not locate public repo root containing config/paths.json")


def load_project_paths(start: Path | None = None) -> dict[str, Path]:
    repo_root = find_repo_root(start)
    config = json.loads((repo_root / "config" / "paths.json").read_text())

    def resolve(value: str | None) -> Path | None:
        if value is None:
            return None
        p = Path(value)
        return p if p.is_absolute() else (repo_root / p)

    outputs_root = resolve(config.get("outputs_root", "outputs"))
    bids_root = resolve(config["bids_root"])

    paths = {
        "repo_root": repo_root,
        "bids_root": bids_root,
        "derivatives_root": bids_root / "derivatives",
        "other_study_data_dir": bids_root / "derivatives" / "01_other_study_data",
        "participants_file": bids_root / "participants.tsv",
        "outputs_root": outputs_root,
        "figures_root": outputs_root / "figures",
        "tables_root": outputs_root / "tables",
        "logs_root": outputs_root / "logs",
        "kinfitr_output_root": outputs_root / "kinfitr",
        "pk_pd_tables_root": outputs_root / "tables" / "pk_pd",
        "pk_pd_figures_root": outputs_root / "figures" / "pk_pd",
        "nca_output_root": outputs_root / "tables" / "pk_pd" / "nca",
        "surface_figures_root": outputs_root / "figures" / "surface_analysis",
        "surface_tables_root": outputs_root / "tables" / "surface_analysis",
        "demographics_tables_root": outputs_root / "tables" / "demographics",
    }
    return paths


def ensure_project_dirs(paths: dict[str, Path] | None = None) -> dict[str, Path]:
    paths = load_project_paths() if paths is None else paths
    for key in (
        "figures_root",
        "tables_root",
        "logs_root",
        "kinfitr_output_root",
        "pk_pd_tables_root",
        "pk_pd_figures_root",
        "nca_output_root",
        "surface_figures_root",
        "surface_tables_root",
        "demographics_tables_root",
    ):
        paths[key].mkdir(parents=True, exist_ok=True)
    return paths

File: code/utils/project_paths.R
find_repo_root <- function(start = getwd()) {
  current <- normalizePath(start, winslash = "/", mustWork = TRUE)
  repeat {
    if (file.exists(file.path(current, "config", "paths.json"))) {
      return(current)
    }
    parent <- dirname(current)
    if (identical(parent, current)) {
      stop("Could not locate repo root containing config/paths.json")
    }
    current <- parent
  }
}

load_project_paths <- function(start = getwd()) {
  repo_root <- find_repo_root(start)
  config <- jsonlite::read_json(file.path(repo_root, "config", "paths.json"), simplifyVector = TRUE)

  resolve_path <- function(path_value) {
    if (is.null(path_value) || identical(path_value, "")) {
      return(NULL)
    }
    if (grepl("^/", path_value)) {
      return(path_value)
    }
    return(file.path(repo_root, path_value))
  }

  outputs_root <- resolve_path(ifelse(is.null(config$outputs_root), "outputs", config$outputs_root))
  bids_root <- resolve_path(config$bids_root)

  paths <- list(
    repo_root = repo_root,
    bids_root = bids_root,
    derivatives_root = file.path(bids_root, "derivatives"),
    other_study_data_dir = file.path(bids_root, "derivatives", "01_other_study_data"),
    participants_file = file.path(bids_root, "participants.tsv"),
    outputs_root = outputs_root,
    figures_root = file.path(outputs_root, "figures"),
    tables_root = file.path(outputs_root, "tables"),
    logs_root = file.path(outputs_root, "logs"),
    kinfitr_output_root = file.path(outputs_root, "kinfitr")
  )

  dir.create(paths$figures_root, recursive = TRUE, showWarnings = FALSE)
  dir.create(paths$tables_root, recursive = TRUE, showWarnings = FALSE)
  dir.create(paths$logs_root, recursive = TRUE, showWarnings = FALSE)
  dir.create(paths$kinfitr_output_root, recursive = TRUE, showWarnings = FALSE)

  paths
}
